### ISSUE TITLE
ZLayer does not support fork

### ISSUE THREAD

**Issue by fillson-shady** [zio/zio#3098]
I have 2 almost identical tests in which `ZLayer` with pool of forked workers and a work `Queue` was created.  
First test written for `zio:1.0.0-RC17` works fine, however second one written for `zio:1.0.0-RC18` got stuck on `queue.take` call and never completes.
  
This is succeeded test (for `zio:1.0.0-RC17`):
```scala
import zio._
import zio.clock._
import zio.console._
import zio.duration._

object Test extends App {

  def fork(poolSize: Int)
          (worker: List[String] => List[String]): ZLayer.NoDeps[Nothing, Has[Queue[List[String]]]] =
    ZLayer.fromEffect { for {
      queue   <- Queue.bounded[List[String]](poolSize)
      workers  = List.fill(poolSize) { for {
        batch    <- queue.take
        results  <- ZIO.effect(worker(batch))
      } yield ()}
      _       <- ZIO.forkAll(workers.map(_.forever))
    } yield Has(queue) }

  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] = (for {
    _ <- putStrLn("sleeping")
    _ <- ZIO.sleep(10.seconds)
    _ <- putStrLn("sending")
    _ <- ZIO.accessM[Has[Queue[List[String]]]](_.get.offer(List("1", "2", "3")))
    _ <- putStrLn("done")
    _ <- ZIO.sleep(Duration.Infinity)
  } yield ()).provideLayer(
    fork(1)(l => l.tapEach(println)) ++ Console.live ++ Clock.live
  ).as(0)
}
```
  
This is failing test (for `zio:1.0.0-RC18`):
```scala
import zio._
import zio.console._
import zio.duration._

object Test extends App {

  def fork(poolSize: Int)
          (worker: List[String] => List[String]) =
    ZLayer.fromEffect { for {
      queue   <- Queue.bounded[List[String]](poolSize)
      workers  = List.fill(poolSize) { for {
        batch    <- queue.take
        results  <- ZIO.effect(worker(batch))
      } yield ()}
      _       <- ZIO.forkAll(workers.map(_.forever))
    } yield queue }

  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] = (for {
    _ <- putStrLn("sleeping")
    _ <- ZIO.sleep(10.seconds)
    _ <- putStrLn("sending")
    _ <- ZIO.accessM[Has[Queue[List[String]]]](_.get.offer(List("1", "2", "3")))
    _ <- putStrLn("done")
    _ <- ZIO.sleep(Duration.Infinity)
  } yield ()).provideCustomLayer(
    fork(1)(l => l.tapEach(println))
  ).as(0)
}
```
  
Is it a bug or `ZLayer` is not designed to be used in such way?

**Comment by adamgfraser** [comment]
@fillson-shady We believe this is related to #3073 since zipWithPar is used internally in ZLayer. We are working on a fix.

**Comment by fillson-shady** [comment]
@adamgfraser Thank you so much for the quick reply and great support! I'm sure it will be fixed in `zio:1.0.0`

**Comment by adamgfraser** [comment]
Resolved by #3829.
